Incorporating frequency visualization with datastore

mnuzen opened the pull request:
# Description

- Incorporated datastore data into frequency visualization by fetching data with DataServlet 
- Added edges on graph based on frequency attribute (number of times connection was made between 2 addresses)
- Demo of visualization on /network.html page 

- Still need to incorporate more testing and modularize

# Checklist:

- [x] Code follows the style guidelines
- [x] Code commented in relevant places
- [x] Changes to necessary documentation made
- [x] New and existing unit tests pass locally with changes
- [x] Basic manual testing performed after deploying app

mnuzen committed:
Removing copied files

mnuzen committed:
Merge branch 'setup' of https://github.com/mnuzen/step-capstone-2020 into setup

mnuzen committed:
Merge branch 'setup' of https://github.com/mnuzen/step-capstone-2020 into setup

mnuzen committed:
Merge branch 'setup' of https://github.com/mnuzen/step-capstone-2020 into setup

Commit:
Added hardcode Json table

mnuzen committed:
Adding network page

mnuzen committed:
Moving network visualizations into code

mnuzen committed:
Adding pcap functionality

Erik-Handeland committed:
Switch from local JSON to dataStore

mnuzen committed:
Adding more functionality

mnuzen committed:
Adding fetch functionality

Erik-Handeland committed:
added flagged option to datastore form

mnuzen committed:
Additional pcap files

mnuzen committed:
Additional pcap files

mnuzen committed:
Multiple pcap files

Erik-Handeland committed:
Quick add button for random table data

Erik-Handeland committed:
replaced mock data array with read from CSV file

mnuzen committed:
Changing network styling

mnuzen committed:
Merge branch 'Erik_Testing' into melba_proto

mnuzen committed:
Merging branches to incorporate datastore

mnuzen committed:
Adding datastore func

Erik-Handeland committed:
chaged datastore to hold freq. attribute

mnuzen committed:
Datastore incorp

mnuzen committed:
Merge remote-tracking branch 'origin/Erik_Testing' into melba_datastore

mnuzen committed:
More datastore func

mnuzen committed:
Adding datastore functionality

mnuzen committed:
Modularizing code

mnuzen committed:
Refactoring code

mnuzen committed:
Commenting code

mnuzen committed:
Removing incomplete packet parser

mnuzen committed:
Refactoring code

mnuzen committed:
Refactoring code again

mnuzen committed:
Frequencies in packet connections to graph

arunk054 commented:
Ah I did not know about Draft pull requests, looks like this is a  new Github feature in the last one year or so.

arunk054 commented:
There are lots of files in your PR ~ 1820, did you rebase your branch with master? You either do a git pull or git fetch + git merge, the latter would create a merge commit for you that would resolve conflicts.

Erik-Handeland commented:
Some of those files might of come from Erik_Testing branch. I did have to add a some of the files for the bootstrap that were missing from main for my table visualization. I ended up adding the `vender` folder into `web app`. Looking at the file structure it seems that a subfolder `fontawesome-free` seems to have over 1,700 files.

mnuzen committed:
Removing vendor folder

mnuzen commented:
Okay, those files have been removed.

mnuzen committed:
Commenting code

Erik-Handeland committed:
Abstracted database upload and retrieval into seperate classes

arunk054 commented on src/main/java/com/google/sps/data/PCAPdata.java:
```diff
@@ -0,0 +1,40 @@
+  String protocol;
+  int size;
+  boolean flagged;
+  int frequency;
```
Make them private or if you dont want to write getters make them protected for now.

arunk054 commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -1,32 +1,94 @@
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
```
You should not be using any of this datastore APIs in your servlet directly. You should be calling Erik's DAOs that expose functions you need. Like you have PCAPDao interface that provides a method getAllPCAPTuples(String id); And Erik could have two implementations one with mock data and one with  real. Your servlet will just need to instantiate one of them at init() depending on whether you are testing with mocks or real. The rest of the servlet is agnostic of that.

mnuzen replied:
Yes, sorry we didn't make that clear -- right now the DataServlet and LoadMockData are Erik's files which I merged into my branch to use. So we can work together to refactor the code

arunk054 commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -1,32 +1,94 @@
+    ArrayList<PCAPdata> dataTable = new ArrayList<PCAPdata>();
+    
+      
+    Query query = new Query("data").addSort("Flagged", SortDirection.DESCENDING); //switch to time when database gets new datafield
```
Is this fetching all the tuples in a. pcap file? I assume you have already parsed the PCAP file and loaded it into the data store.  And I believe you only have one file. How do you differentiate the data if there are multiple files parsed and uploaded?

mnuzen replied:
We were going to have different IDs for each file; so right now all of our data is tagged with "data" but we could tag it with "file1" or "file2" and then retrieve only the data with those tags.

arunk054 commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -1,32 +1,94 @@
+      int frequency = (int) (long) entity.getProperty("Frequency");
+
+      PCAPdata temp = new PCAPdata(source, destination, domain, 
+      location,protocol, size, flagged, frequency);
```
All of this can go in a separate function where you just return a PCAP data object.
`CreatePCAPData(entity)`

arunk054 commented on src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -1,32 +1,94 @@
+        boolean flagged = Boolean.parseBoolean(request.getParameter("flagged"));
+        int frequency = Integer.parseInt(request.getParameter("frequency"));
+
+        Entity pcapEntity = new Entity("data"); //creates entitiy tab in datastore
```
As i pointed before this should be in your servlet, you should call some other function that is provided to you by the datastore.  I guess that is still in development.

arunk054 commented on src/main/java/com/google/sps/servlets/LoadMockData.java:
```diff
@@ -0,0 +1,70 @@
+
+import java.io.*; 
+
+@WebServlet("/load_data")
```
make it explicit like load_mock_data

arunk054 commented on src/main/java/com/google/sps/servlets/LoadMockData.java:
```diff
@@ -0,0 +1,70 @@
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+
+      //CSV format: Source,Destination,Domain,Location,Protocal,Size,Flagged,Frequency
+        String csvFile = "data.csv"; //CSV located in project dir /webapp
+        String line = "";
+        String cvsSplitBy = ",";
+
+        try (BufferedReader br = new BufferedReader(new FileReader(csvFile))) {
+
+            while ((line = br.readLine()) != null) {
+
+                // use comma as separator
+                String[] pcapLine = line.split(cvsSplitBy);
+
+                Entity pcapEntity = new Entity("data");
+                pcapEntity.setProperty("Source", pcapLine[0]);
+                pcapEntity.setProperty("Destination", pcapLine[1]);
+                pcapEntity.setProperty("Domain", pcapLine[2]);
+                pcapEntity.setProperty("Location", pcapLine[3]);
+                pcapEntity.setProperty("Protocol", pcapLine[4]);
+                pcapEntity.setProperty("Size", Integer.parseInt(pcapLine[5]));
+                pcapEntity.setProperty("Flagged", Boolean.parseBoolean(pcapLine[6]));
+                pcapEntity.setProperty("Frequency", Integer.parseInt(pcapLine[7]));
+                datastore.put(pcapEntity); //pushes new entry to datastore
+            }
+
```
Create a separate class called `MockDataLoader` That class has a function `loadData(List<PCAPData> pcapObjs)`  make that class Abstract by having an abstract method `createPCAPObjects()`. Now you can extend the MockDataLoader with `CSVMockDataLoader` in which you implement the `createPCAPObjects` by reading your csv file and creating PCAP objects out of them, and store them in the class as a private list. THe `loadData` has a generic implementation where you iterate the PCAP objects and call Erik's APIs to add them to data store.

arunk054 commented on src/main/webapp/js/network.js:
```diff
@@ -0,0 +1,173 @@
+  /** Populates graph with hard-coded data. */
+  function populateGraph(){
+    nodes = [
+      { id: 0, label: "My Computer", group: 0 },
+      { id: 4, label: "Class C", group: 6 },
+      { id: 7, label: "Class A", group: 8 },
+      { id: 13, label: "Class E", group: 7 },
+      { id: 15, label: "Class D", group: 5 },
+      { id: 16, label: "224.0.0.0", group: 5 },
+      { id: 17, label: "224.0.0.1", group: 5 },
+      { id: 18, label: "193.0.1.1", group: 6 },
+      { id: 19, label: "192.0.1.1", group: 6 },
+      { id: 21, label: "240.0.0.0", group: 7 },
+      { id: 22, label: "241.0.0.0", group: 7 },
+      { id: 23, label: "242.0.0.0", group: 7 },
+      { id: 24, label: "1.0.0.1", group: 8 },
+      { id: 25, label: "1.1.1.1", group: 8 },
+      { id: 26, label: "126.1.1.1", group: 8 },
+      { id: 27, label: "128.2.0.1", group: 9 },
+      { id: 28, label: "Class B", group: 9 },
+      { id: 29, label: "128.1.0.1", group: 9 },
+    ];
```
I am not sure how you are going to do this grouping with real data. I believe you have some ideas?

Erik-Handeland committed:
removed vendor folder

mnuzen committed:
Incorporating Erik's changes

Erik-Handeland committed:
Combind get/set Datastore

Combind seperate datastore classes and reformated code to match google style guide.

mnuzen committed:
Added hard-coded groupings

mnuzen committed:
Merge remote-tracking branch 'origin/Erik_Testing' into melba_datastore

mnuzen committed:
Refactoring code

mnuzen committed:
Modifying network graph physics

mnuzen committed:
Pulling out force constants

mnuzen committed:
Changing chart sizing

arunk054 commented on src/main/webapp/js/network.js:
```diff
@@ -0,0 +1,128 @@
+        nodes[node] = {id: node, label: data[i].destination, group: node};
+
+        
+        if (ip == 17){ 
```
Just a add a comment of why you have this hack?

arunk054 commented on src/main/java/com/google/sps/datastore/GenericPCAPDao.java:
```diff
@@ -0,0 +1,66 @@
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+
+public class GenericPCAPDao {
```
Just create an interface GenericPCAPDao and change this to GenericPCAPDaoImpl

arunk054 commented on src/main/java/com/google/sps/servlets/LoadMockData.java:
```diff
@@ -0,0 +1,56 @@
+             Integer.parseInt(pcapLine[7]));
+
+        GenericPCAPDao data = new GenericPCAPDao();
+        data.setPCAPObjects(tempPCAP, "file_1");
```
make sure to create const strings instread of hardcoding directly. and add a todo or pass the entity name / file name as a parameter.

arunk054 commented on src/main/java/com/google/sps/servlets/TableServlet.java:
```diff
@@ -0,0 +1,60 @@
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    GenericPCAPDao data = new GenericPCAPDao();
+
+    String json = convertToJsonUsingGson(data.getPCAPObjects("file_1"));
```
Again avoid hardcoding this, pass it as a param `PCAPId`

arunk054 reviewed (approved):
Looks very good.

Erik-Handeland committed:
Const for filename

Erik-Handeland committed:
Added datastore interface

Erik-Handeland committed:
Seperated mockData servlet into seperate MockDataLoader class

Erik-Handeland committed:
Update data.csv

mnuzen committed:
Incorporating latest datastore changes

mnuzen committed:
Modifications to csv data

mnuzen committed:
Adding navigation buttons

mnuzen committed:
Commenting code

mnuzen commented:
- Incorporated new changes to datastore
- Added navigation buttons on top of network visualization page
- Commented on hard-coded groups of IP addresses (based on first two bits of address) for now

Erik-Handeland committed:
modified CSV loader

mnuzen committed:
Merge remote-tracking branch 'origin/Erik_Testing' into melba_datastore
